File: deploy/smoke/cpp/infer.cpp
#include <gflags/gflags.h>
#include <glog/logging.h>

#include <iostream>
#include <numeric>

#include "opencv2/core.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/imgproc.hpp"
#include "paddle/include/paddle_inference_api.h"

using paddle_infer::Config;
using paddle_infer::CreatePredictor;
using paddle_infer::PrecisionType;
using paddle_infer::Predictor;

DEFINE_string(model_file, "", "Directory of the inference model.");
DEFINE_string(params_file, "", "Directory of the inference model.");
DEFINE_string(image, "", "Image path to be predicted.");
DEFINE_bool(use_gpu, false, "Whether to use gpu.");
DEFINE_bool(use_trt, false, "Whether to use trt.");
DEFINE_int32(trt_precision, 0,
             "Precision type of tensorrt, 0: kFloat32, 1: kInt8, 2: kHalf");
DEFINE_bool(collect_dynamic_shape_info, false,
            "Whether to collect dynamic shape before using tensorrt");
DEFINE_string(dynamic_shape_file, "dynamic_shape_info.txt",
              "Path of a dynamic shape file for tensorrt");

void get_image(const std::string &image, float *data) {
  cv::Mat img = cv::imread(image, cv::IMREAD_COLOR);

  cv::cvtColor(img, img, cv::COLOR_BGR2RGB);
  cv::resize(img, img, cv::Size(1280, 384));
  // Normalize
  img.convertTo(img, CV_32F, 1.0 / 255, 0);

  std::vector<float> mean_values{0.485, 0.456, 0.406};
  std::vector<float> std_values{0.229, 0.224, 0.225};

  std::vector<cv::Mat> rgbChannels(3);
  cv::split(img, rgbChannels);

  for (int i = 0; i < 3; ++i) {
    rgbChannels[i].convertTo(rgbChannels[i], CV_32FC1, 1 / std_values[i],
                             (0.0 - mean_values[i]) / std_values[i]);
  }

  cv::merge(rgbChannels, img);

  // from hwc to chw
  int rows = img.rows;
  int cols = img.cols;
  int chs = img.channels();

  for (int i = 0; i < chs; ++i) {
    cv::extractChannel(
        img, cv::Mat(rows, cols, CV_32FC1, data + i * rows * cols), i);
  }
}

std::shared_ptr<Predictor> InitPredictor() {
  Config config;
  config.SetModel(FLAGS_model_file, FLAGS_params_file);

  if (FLAGS_use_gpu) {
    config.EnableUseGpu(1000, 0);
  } else {
    config.EnableMKLDNN();
  }

  if (FLAGS_collect_dynamic_shape_info) {
    config.CollectShapeRangeInfo(FLAGS_dynamic_shape_file);
  } else if (FLAGS_use_trt) {
    config.EnableTensorRtEngine(
        1 << 30, 1, 5, PrecisionType(FLAGS_trt_precision), false, false);
    config.EnableTunedTensorRtDynamicShape(FLAGS_dynamic_shape_file, true);
  }

  config.SwitchIrOptim(true);
  config.EnableMemoryOptim();

  return CreatePredictor(config);
}

void run(Predictor *predictor, const std::vector<float> &input_im_data,
         const std::vector<int> &input_im_shape,
         const std::vector<float> &input_K_data,
         const std::vector<int> &input_K_shape,
         const std::vector<float> &input_ratio_data,
         const std::vector<int> &input_ratio_shape,
         std::vector<float> &out_data) {
  int input_num = std::accumulate(input_im_shape.begin(), input_im_shape.end(),
                                  1, std::multiplies<int>());

  auto input_names = predictor->GetInputNames();
  auto output_names = predictor->GetOutputNames();

  auto input_im_handle = predictor->GetInputHandle(input_names[0]);
  input_im_handle->Reshape(input_im_shape);
  input_im_handle->CopyFromCpu(input_im_data.data());

  auto input_K_handle = predictor->GetInputHandle(input_names[1]);
  input_K_handle->Reshape(input_K_shape);
  input_K_handle->CopyFromCpu(input_K_data.data());

  auto input_ratio_handle = predictor->GetInputHandle(input_names[2]);
  input_ratio_handle->Reshape(input_ratio_shape);
  input_ratio_handle->CopyFromCpu(input_ratio_data.data());

  CHECK(predictor->Run());
  auto output_t = predictor->GetOutputHandle(output_names[0]);

  auto outshape = output_t->shape();
  int outsize = std::accumulate(outshape.begin(), outshape.end(), 1,
                                std::multiplies<int>());
  out_data.resize(outsize);

  output_t->CopyToCpu(out_data.data());
}

int main(int argc, char *argv[]) {
  google::ParseCommandLineFlags(&argc, &argv, true);
  if (FLAGS_model_file == "" || FLAGS_params_file == "" || FLAGS_image == "") {
    std::cout << "Missing required parameter" << std::endl;
    std::cout << "Usage: " << std::string(argv[0])
              << " --model_file ${MODEL_FILE}"
              << "--params_file ${PARAMS_FILE} "
              << "--image ${TEST_IMAGE}" << std::endl;
    return -1;
  }

  auto predictor = InitPredictor();

  std::vector<int> input_im_shape = {1, 3, 384, 1280};
  std::vector<float> input_im_data(1 * 3 * 384 * 1280);
  get_image(FLAGS_image, input_im_data.data());

  std::vector<int> input_K_shape = {1, 3, 3};
  // Listed below are camera intrinsic parameter of the kitti dataset
  // If the model is trained on other datasets, please replace the relevant data
  std::vector<float> input_K_data = {
      721.53771973, 0., 609.55932617, 0., 721.53771973, 172.85400391, 0, 0, 1};
  std::vector<int> input_ratio_shape = {1, 2};
  std::vector<float> input_ratio_data(4, 4);

  std::vector<float> out_data;
  run(predictor.get(), input_im_data, input_im_shape, input_K_data,
      input_K_shape, input_ratio_data, input_ratio_shape, out_data);

  std::vector<std::vector<float>> results;
  for (int i = 0; i < out_data.size(); i += 14) {
    // item 1       :  class
    // item 2       :  observation angle α
    // item 3 ~ 6   :  box2d x1, y1, x2, y2
    // item 7 ~ 9   :  box3d h, w, l
    // item 10 ~ 12 :  box3d bottom center x, y, z
    // item 13      :  box3d yaw angle
    // item 14      :  score
    std::vector<float> vec(out_data.begin() + i, out_data.begin() + i + 14);
    results.push_back(vec);
  }

  for (const auto &res : results) {
    // Filter predictions with low scores
    if (res[13] <= 0.25) continue;
    for (const auto &item : res) {
      std::cout << item << " ";
    }
    std::cout << std::endl;
  }
}

File: deploy/smoke/python/infer.py
# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse

import cv2
import numpy as np
import paddle
from paddle.inference import Config, PrecisionType, create_predictor


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--model_file",
        type=str,
        default="./inference.pdmodel",
        help="Model filename, Specify this when your model is a combined model."
    )
    parser.add_argument(
        "--params_file",
        type=str,
        default="./inference.pdiparams",
        help=
        "Parameter filename, Specify this when your model is a combined model.")
    parser.add_argument(
        '--image', dest='image', help='The image path', type=str, required=True)
    parser.add_argument(
        "--use_gpu", action='store_true', help="Whether use gpu.")
    parser.add_argument(
        "--use_trt", action='store_true', help="Whether use trt.")
    parser.add_argument(
        "--collect_dynamic_shape_info",
        action='store_true',
        help="Whether to collect dynamic shape before using tensorrt.")
    parser.add_argument(
        "--dynamic_shape_file",
        dest='dynamic_shape_file',
        help='The image path',
        type=str,
        default="dynamic_shape_info.txt")
    return parser.parse_args()


def get_ratio(ori_img_size, output_size, down_ratio=(4, 4)):
    return np.array([[
        down_ratio[1] * ori_img_size[1] / output_size[1],
        down_ratio[0] * ori_img_size[0] / output_size[0]
    ]], np.float32)


def get_img(img_path):
    img = cv2.imread(img_path)
    origin_shape = img.shape
    img = cv2.resize(img, (1280, 384))

    target_shape = img.shape
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)

    img = img / 255.0
    img = np.subtract(img, np.array([0.485, 0.456, 0.406]))
    img = np.true_divide(img, np.array([0.229, 0.224, 0.225]))
    img = np.array(img, np.float32)

    img = img.transpose(2, 0, 1)
    img = img[None, :, :, :]

    return img, origin_shape, target_shape


def init_predictor(args):
    config = Config(args.model_file, args.params_file)
    config.enable_memory_optim()
    if args.use_gpu:
        config.enable_use_gpu(1000, 0)
    else:
        # If not specific mkldnn, you can set the blas thread.
        # The thread num should not be greater than the number of cores in the CPU.
        config.set_cpu_math_library_num_threads(4)
        config.enable_mkldnn()

    if args.collect_dynamic_shape_info:
        config.collect_shape_range_info(args.dynamic_shape_file)
    elif args.use_trt:
        allow_build_at_runtime = True
        config.enable_tuned_tensorrt_dynamic_shape(args.dynamic_shape_file,
                                                   allow_build_at_runtime)

        config.enable_tensorrt_engine(
            workspace_size=1 << 20,
            max_batch_size=1,
            min_subgraph_size=3,
            precision_mode=PrecisionType.Float32)

    predictor = create_predictor(config)
    return predictor


def run(predictor, image, K, down_ratio):
    # copy img data to input tensor
    input_names = predictor.get_input_names()
    for i, name in enumerate(input_names):
        input_tensor = predictor.get_input_handle(name)
        if name == "images":
            input_tensor.reshape(image.shape)
            input_tensor.copy_from_cpu(image.copy())
        elif name == "trans_cam_to_img":
            input_tensor.reshape(K.shape)
            input_tensor.copy_from_cpu(K.copy())
        elif name == "down_ratios":
            input_tensor.reshape(down_ratio.shape)
            input_tensor.copy_from_cpu(down_ratio.copy())

    # do the inference
    predictor.run()

    results = []
    # get out data from output tensor
    output_names = predictor.get_output_names()
    for i, name in enumerate(output_names):
        output_tensor = predictor.get_output_handle(name)
        output_data = output_tensor.copy_to_cpu()
        results.append(output_data)

    return results


if __name__ == '__main__':
    args = parse_args()
    pred = init_predictor(args)
    # Listed below are camera intrinsic parameter of the kitti dataset
    # If the model is trained on other datasets, please replace the relevant data
    K = np.array([[[721.53771973, 0., 609.55932617],
                   [0., 721.53771973, 172.85400391], [0, 0, 1]]], np.float32)

    img, ori_img_size, output_size = get_img(args.image)
    ratio = get_ratio(ori_img_size, output_size)

    results = run(pred, img, K, ratio)

    total_pred = results[0]
